File: languages/csharp/AvalonSdk/Social.cs
// Friends and presence — capability-gated reads/writes on Session,
// mirroring crates/sdk/src/social.rs.
//
// Every method here checks its own capability *before* making any request,
// same convention the Rust module established — a Session with no grants
// rejects without ever touching the network.
//
// update_presence/presence_of/subscribe_presence match the Rust SDK's own
// documented gaps rather than inventing stricter behavior: no visibility
// scoping on presence_of, and update_presence is an identity
// publishing its own status, not an integrator-authority publish (no
// IntegratorCredential/IntegratorBinding capability-grant system exists yet).

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Channels;
using System.Threading.Tasks;

namespace Avalon.Sdk
{
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum PresenceStatus
    {
        Online,
        Away,
        DoNotDisturb,
        Offline,
    }

    public sealed class Presence
    {
        [JsonPropertyName("identity_id")]
        public Guid IdentityId { get; set; }

        [JsonPropertyName("status")]
        public PresenceStatus Status { get; set; }

        /// <summary>The integrator the user is currently active in, if any and if shared.</summary>
        [JsonPropertyName("active_in")]
        public Guid? ActiveIn { get; set; }

        [JsonPropertyName("updated_at")]
        public DateTimeOffset UpdatedAt { get; set; }
    }

    /// <summary>
    /// A friend, from this integrator's point of view. DisplayName is always null today —
    /// GET /friends returns only the two identity ids and the friendship's since timestamp,
    /// no profile-lookup endpoint exists yet to resolve it (matches Friend's Rust doc comment).
    /// </summary>
    public sealed class Friend
    {
        public Friend(Guid identityId, string? displayName, Presence? presence)
        {
            IdentityId = identityId;
            DisplayName = displayName;
            Presence = presence;
        }

        public Guid IdentityId { get; }
        public string? DisplayName { get; }

        /// <summary>Populated only when presence.read is also granted alongside friends.read.</summary>
        public Presence? Presence { get; }
    }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    internal enum PresenceSubscribeMessageType
    {
        Subscribe,
    }

    internal sealed class PresenceSubscribeMessage
    {
        [JsonPropertyName("type")]
        public string Type => "subscribe";

        [JsonPropertyName("ids")]
        public List<Guid> Ids { get; set; } = new List<Guid>();
    }

    public sealed partial class Session
    {
        private static readonly JsonSerializerOptions JsonOptions = new JsonSerializerOptions()
        {
            PropertyNameCaseInsensitive = true,
            // Issue #725: Generated.cs's enums need this — see
            // EnumMemberJsonConverter.cs's own header comment for why.
            Converters = { new EnumMemberJsonConverterFactory() },
        };

        /// <summary>Builds a Friend view from a raw friendship plus whatever presence data is
        /// available for the other party — testable without any HTTP call.</summary>
        private static Friend MergeFriend(Avalon.Sdk.Generated.FriendshipResponse friendship, Guid selfId, IReadOnlyDictionary<Guid, Presence> presenceById)
        {
            var other = friendship.A == selfId ? friendship.B : friendship.A;
            presenceById.TryGetValue(other, out var presence);
            return new Friend(other, null, presence);
        }

        /// <summary>Generated.cs's own <see cref="Avalon.Sdk.Generated.PresenceStatus"/> and
        /// this SDK's public <see cref="PresenceStatus"/> are deliberately two separate enum
        /// types with identical member names, mirroring the <c>ToDomainGenre</c> pattern in
        /// <c>AccountSession.cs</c> — a name round-trip via <see cref="Enum.Parse"/> rather
        /// than a fragile numeric cast.</summary>
        private static PresenceStatus ToDomainPresenceStatus(Avalon.Sdk.Generated.PresenceStatus generated) =>
            (PresenceStatus)Enum.Parse(typeof(PresenceStatus), generated.ToString());

        private static Avalon.Sdk.Generated.PresenceStatus ToGeneratedPresenceStatus(PresenceStatus status) =>
            (Avalon.Sdk.Generated.PresenceStatus)Enum.Parse(typeof(Avalon.Sdk.Generated.PresenceStatus), status.ToString());

        /// <summary>Maps a wire-shape <see cref="Avalon.Sdk.Generated.PresenceResponse"/>
        /// (used for the two HTTP presence endpoints) onto this SDK's public
        /// <see cref="Presence"/> — the raw websocket push path in
        /// <see cref="SubscribePresenceAsync"/> deserializes directly into
        /// <see cref="Presence"/> instead, since it has no OpenAPI coverage at all.</summary>
        private static Presence ToDomainPresence(Avalon.Sdk.Generated.PresenceResponse p) => new Presence
        {
            IdentityId = p.IdentityId,
            Status = ToDomainPresenceStatus(p.Status),
            ActiveIn = p.ActiveIn,
            UpdatedAt = p.UpdatedAt,
        };

        /// <summary>server_url is http(s)://…; the websocket endpoint needs ws(s)://….</summary>
        private static string WebSocketUrl(string serverUrl, string path)
        {
            if (serverUrl.StartsWith("https://", StringComparison.Ordinal))
            {
                return "wss://" + serverUrl.Substring("https://".Length) + path;
            }
            if (serverUrl.StartsWith("http://", StringComparison.Ordinal))
            {
                return "ws://" + serverUrl.Substring("http://".Length) + path;
            }
            return serverUrl + path;
        }

        /// <summary>GET /friends — requires friends.read. Also embeds each friend's Presence
        /// when presence.read is granted too, via one batched PresenceOfAsync call.</summary>
        public async Task<IReadOnlyList<Friend>> FriendsAsync(CancellationToken ct = default)
        {
            Require("friends.read");

            using var request = new HttpRequestMessage(HttpMethod.Get, $"{ServerUrl}/friends");
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", Token);
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
            var friendships = await ReadJsonAsync<List<Avalon.Sdk.Generated.FriendshipResponse>>(response, ct).ConfigureAwait(false)
                ?? new List<Avalon.Sdk.Generated.FriendshipResponse>();

            var presenceById = new Dictionary<Guid, Presence>();
            if (HasCapability("presence.read") && friendships.Count > 0)
            {
                var otherIds = friendships.Select(f => f.A == IdentityGuid ? f.B : f.A).ToArray();
                foreach (var presence in await PresenceOfAsync(otherIds, ct).ConfigureAwait(false))
                {
                    presenceById[presence.IdentityId] = presence;
                }
            }

            return friendships.Select(f => MergeFriend(f, IdentityGuid, presenceById)).ToList();
        }

        /// <summary>The calling user's own presence, as the server currently has it.</summary>
        public async Task<Presence> PresenceAsync(CancellationToken ct = default)
        {
            Require("presence.read");
            var mine = await PresenceOfAsync(new[] { IdentityGuid }, ct).ConfigureAwait(false);
            // GET /presence?ids=<one id> always returns exactly one entry — the store always
            // answers for any id (missing/stale reads as Offline).
            return mine[0];
        }

        /// <summary>GET /presence?ids=… for the given identities. No visibility filtering
        /// yet — see the module comment.</summary>
        public async Task<IReadOnlyList<Presence>> PresenceOfAsync(IReadOnlyList<Guid> ids, CancellationToken ct = default)
        {
            Require("presence.read");
            if (ids.Count == 0)
            {
                return Array.Empty<Presence>();
            }

            var idsParam = string.Join(",", ids);
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{ServerUrl}/presence?ids={Uri.EscapeDataString(idsParam)}");
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", Token);
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
            var presences = await ReadJsonAsync<List<Avalon.Sdk.Generated.PresenceResponse>>(response, ct).ConfigureAwait(false)
                ?? new List<Avalon.Sdk.Generated.PresenceResponse>();
            return presences.Select(ToDomainPresence).ToList();
        }

        /// <summary>PUT /me/presence — an identity publishing their own status. Not
        /// capability-gated: the server requires only a valid identity session for this.</summary>
        public async Task UpdatePresenceAsync(PresenceStatus status, CancellationToken ct = default)
        {
            using var request = new HttpRequestMessage(HttpMethod.Put, $"{ServerUrl}/me/presence");
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", Token);
            request.Content = JsonContent(new Avalon.Sdk.Generated.UpdatePresenceRequest { Status = ToGeneratedPresenceStatus(status) });
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
        }

        /// <summary>
        /// Subscribes to live presence updates for the given ids, additive to
        /// PresenceOfAsync's point-in-time reads. Connects to GET /ws/presence (auth via a
        /// ?token= query parameter — a websocket handshake can't carry a bearer header), sends
        /// one subscribe message, then forwards every pushed Presence into the returned
        /// channel reader on a background task. Disposing the returned CancellationTokenSource
        /// (or cancelling ct) ends the background task; there is no separate unsubscribe call.
        /// </summary>
        public async Task<ChannelReader<Presence>> SubscribePresenceAsync(IReadOnlyList<Guid> ids, CancellationToken ct = default)
        {
            Require("presence.read");

            var url = WebSocketUrl(ServerUrl, $"/ws/presence?token={Uri.EscapeDataString(Token)}");
            var socket = new ClientWebSocket();
            try
            {
                await socket.ConnectAsync(new Uri(url), ct).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                socket.Dispose();
                throw new AvalonWebSocketException(e.Message);
            }

            var subscribe = JsonSerializer.Serialize(new PresenceSubscribeMessage { Ids = ids.ToList() });
            var subscribeBytes = Encoding.UTF8.GetBytes(subscribe);
            try
            {
                await socket.SendAsync(new ArraySegment<byte>(subscribeBytes), WebSocketMessageType.Text, true, ct)
                    .ConfigureAwait(false);
            }
            catch (Exception e)
            {
                socket.Dispose();
                throw new AvalonWebSocketException(e.Message);
            }

            var channel = Channel.CreateUnbounded<Presence>();
            _ = Task.Run(async () =>
            {
                var buffer = new byte[8192];
                try
                {
                    while (socket.State == WebSocketState.Open)
                    {
                        using var ms = new System.IO.MemoryStream();
                        WebSocketReceiveResult result;
                        do
                        {
                            result = await socket.ReceiveAsync(new ArraySegment<byte>(buffer), ct).ConfigureAwait(false);
                            if (result.MessageType == WebSocketMessageType.Close)
                            {
                                channel.Writer.TryComplete();
                                return;
                            }
                            ms.Write(buffer, 0, result.Count);
                        } while (!result.EndOfMessage);

                        Presence? presence;
                        try
                        {
                            presence = JsonSerializer.Deserialize<Presence>(ms.ToArray(), JsonOptions);
                        }
                        catch (JsonException)
                        {
                            continue;
                        }
                        if (presence == null)
                        {
                            continue;
                        }
                        if (!channel.Writer.TryWrite(presence))
                        {
                            break;
                        }
                    }
                }
                catch (OperationCanceledException)
                {
                    // ct cancelled — end the background task quietly, same as dropping the
                    // Rust receiver ending its forwarding task.
                }
                catch (Exception)
                {
                    // Any other socket failure just ends delivery; the caller observes this as
                    // the channel completing with nothing further.
                }
                finally
                {
                    channel.Writer.TryComplete();
                    socket.Dispose();
                }
            }, ct);

            return channel.Reader;
        }

        /// <summary>PUT /presence/{identity_id} — an <i>integrator</i> setting
        /// presence on behalf of an identity within a capability grant (distinct from
        /// <see cref="UpdatePresenceAsync"/>, which is the identity publishing its own status
        /// directly). Requires presence.publish and this session's own configured
        /// <see cref="IntegratorSlug"/>/<see cref="SigningKey"/> — authenticated the same
        /// challenge-response way <see cref="IssueAchievementAsync"/> is, plus the
        /// <c>x-avalon-identity-id</c> header identifying whose presence is being set; the
        /// server rejects a mismatch between that header and <paramref name="identityId"/>
        /// itself, not just a missing grant.</summary>
        public async Task<Presence> UpdateIntegratorPresenceAsync(
            Guid identityId, PresenceStatus status, Guid? activeIn = null, CancellationToken ct = default)
        {
            Require("presence.publish");

            if (IntegratorSlug is null || SigningKey is null)
            {
                throw new MissingIssuerCredentialsException();
            }

            using var request = new HttpRequestMessage(HttpMethod.Put, $"{ServerUrl}/presence/{identityId}");
            await AttachIntegratorAuthAsync(request, ct).ConfigureAwait(false);
            request.Headers.Add("x-avalon-identity-id", identityId.ToString());
            request.Content = JsonContent(new Avalon.Sdk.Generated.UpdateIntegratorPresenceRequest
            {
                Status = ToGeneratedPresenceStatus(status),
                ActiveIn = activeIn,
            });

            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
            var body = await ReadJsonAsync<Avalon.Sdk.Generated.PresenceResponse>(response, ct).ConfigureAwait(false);
            return ToDomainPresence(body);
        }

        internal static async Task<T> ReadJsonAsync<T>(HttpResponseMessage response, CancellationToken ct) where T : class
        {
#if NET5_0_OR_GREATER
            var stream = await response.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
#else
            var stream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
#endif
            var value = await JsonSerializer.DeserializeAsync<T>(stream, JsonOptions, ct).ConfigureAwait(false);
            return value ?? throw new System.Text.Json.JsonException("expected a JSON value, got null");
        }

        private static HttpContent JsonContent<T>(T value)
        {
            var json = JsonSerializer.Serialize(value);
            return new StringContent(json, Encoding.UTF8, "application/json");
        }
    }
}

File: languages/rust/src/social.rs
//! Friends and presence — capability-gated reads/writes on [`crate::Session`]
//! (issue #17).
//!
//! Every method here calls `Session::require` with its exact
//! capability string *before* making any request — a `Session` with no
//! grants (which is every `Session` today; see `crate::AvalonClient::authenticate`)
//! correctly rejects without ever touching the network. The server enforces
//! the same capabilities again once #26–#28 land; this check is a
//! convenience for integrator developers, not the security boundary.
//!
//! ## `update_presence` deviates from issue #17's original design
//!
//! The issue describes presence *publishing* as `AvalonClient::publish_presence(identity_id,
//! status)`, gated on an integrator credential and an active `IntegratorBinding` (#83).
//! What #16 actually built is `PUT /me/presence`: an *identity*, under their own
//! session, publishing their own status — no `active_in` field, can't target
//! another identity. [`Session::update_presence`] wraps that. #26/#28/#83 have
//! since landed a real integrator-authority path too — `PUT
//! /presence/{identity_id}`, wrapped by [`Session::update_integrator_presence`]
//! (closed out by #741/#749): an integrator, under its own challenge-response
//! credential plus an active `presence.publish` grant, publishing on behalf of
//! this session's own identity, `active_in` included.
//!
//! ## `presence_of` has no visibility filtering yet
//!
//! The server performs no friends/guild/private scoping on `GET /presence`
//! (deferred to #87) — any valid session can look up presence for any ids it
//! names. This method returns exactly what the server returns; it does not
//! (and cannot) narrow that further on the client side, since narrowing
//! client-side would just be security theater on top of a server that
//! doesn't enforce it.

use std::collections::HashMap;

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;
use ed25519_dalek::{Signer, SigningKey};

use crate::types::ids::IdentityId;
use crate::types::permissions::Capability;
use crate::types::social::{Friendship, Presence, PresenceStatus};
use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use tokio_tungstenite::tungstenite::Message as WsMessage;
use uuid::Uuid;

use crate::http::websocket_url;
use crate::{SdkError, Session};

/// A friend, from this integrator's point of view.
///
/// `display_name` is always `None` today: `GET /friends` returns only the
/// two identity ids and the friendship's `since` timestamp (see
/// `crates/server/src/friends.rs`) — no endpoint resolves another identity's
/// profile yet. Filling this in needs a profile-lookup endpoint, which
/// doesn't exist; tracked as a documented gap here rather than guessed at.
#[derive(Debug, Clone)]
pub struct Friend {
    /// The friend's own identity id.
    pub identity_id: IdentityId,
    /// Always `None` today — see this struct's own doc comment.
    pub display_name: Option<String>,
    /// Populated only if `presence.read` is also granted alongside
    /// `friends.read` — otherwise always `None`, so an integrator with only
    /// `friends.read` gets names (once resolvable) and nothing else.
    pub presence: Option<Presence>,
}

/// Builds a [`Friend`] view from a raw [`Friendship`] plus whatever presence
/// data is available for the *other* party. Kept as a free function, testable
/// without any HTTP call: an empty `presence_by_id` (which is exactly what
/// callers pass when `presence.read` isn't granted) always yields
/// `presence: None`.
fn merge_friend(
    friendship: &Friendship,
    self_id: IdentityId,
    presence_by_id: &HashMap<IdentityId, Presence>,
) -> Friend {
    let other = if friendship.a == self_id {
        friendship.b
    } else {
        friendship.a
    };
    Friend {
        identity_id: other,
        display_name: None,
        presence: presence_by_id.get(&other).cloned(),
    }
}

#[derive(Serialize)]
struct UpdatePresenceRequest {
    status: PresenceStatus,
}

/// What `presence::presence_ws` (`crates/server/src/presence.rs`, issue
/// #136) accepts from a connected client. Only one variant exists today;
/// kept as a tagged enum (`{"type":"subscribe","ids":[...]}`) rather than a
/// bare struct so a second client-to-server message kind can be added later
/// without a wire-format break.
#[derive(Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum PresenceSubscribeMessage {
    Subscribe { ids: Vec<IdentityId> },
}

impl Session {
    /// `GET /friends` — requires `friends.read`. Also embeds each friend's
    /// [`Presence`] when `presence.read` is granted too, via a single batched
    /// `presence_of` call rather than one request per friend.
    pub async fn friends(&self) -> Result<Vec<Friend>, SdkError> {
        self.require(Capability::FriendsRead)?;

        let response = crate::http::send(&self.http, &self.retry, true, |c| {
            c.get(format!("{}/friends", self.server_url))
                .bearer_auth(&self.token)
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        let friendships: Vec<Friendship> = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;

        let presence_by_id =
            if self.require(Capability::PresenceRead).is_ok() && !friendships.is_empty() {
                let other_ids: Vec<IdentityId> = friendships
                    .iter()
                    .map(|f| if f.a == self.identity().id { f.b } else { f.a })
                    .collect();
                self.presence_of(&other_ids)
                    .await?
                    .into_iter()
                    .map(|p| (p.identity_id, p))
                    .collect()
            } else {
                HashMap::new()
            };

        Ok(friendships
            .iter()
            .map(|f| merge_friend(f, self.identity().id, &presence_by_id))
            .collect())
    }

    /// The calling user's own presence, as the server currently has it.
    /// Requires `presence.read`.
    pub async fn presence(&self) -> Result<Presence, SdkError> {
        self.require(Capability::PresenceRead)?;
        let mine = self.presence_of(&[self.identity().id]).await?;
        // The store always answers for any id (missing/stale reads as
        // Offline — see crates/server/src/presence.rs), so this is always
        // populated; `expect` documents that invariant rather than masking
        // a real failure behind a made-up default.
        Ok(mine
            .into_iter()
            .next()
            .expect("GET /presence?ids=<one id> always returns exactly one entry"))
    }

    /// `GET /presence?ids=…` for the given identities. Requires
    /// `presence.read`. No visibility filtering — see module docs.
    pub async fn presence_of(&self, ids: &[IdentityId]) -> Result<Vec<Presence>, SdkError> {
        self.require(Capability::PresenceRead)?;
        if ids.is_empty() {
            return Ok(Vec::new());
        }

        let ids_param = ids
            .iter()
            .map(|id| id.0.to_string())
            .collect::<Vec<_>>()
            .join(",");
        let response = crate::http::send(&self.http, &self.retry, true, |c| {
            c.get(format!("{}/presence", self.server_url))
                .query(&[("ids", ids_param.clone())])
                .bearer_auth(&self.token)
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))
    }

    /// `PUT /me/presence` — an identity publishing their own status. See the
    /// module-level docs for why this lives here rather than as
    /// `AvalonClient::publish_presence`. Not capability-gated: the server
    /// requires only a valid identity session for this, matching what's
    /// implemented, not the integrator-credential design #17 originally
    /// described.
    pub async fn update_presence(&self, status: PresenceStatus) -> Result<(), SdkError> {
        // `PUT` is idempotent by its own HTTP semantics — replacing "my
        // status" with the same value twice is safe with no separate
        // dedup key needed, unlike `POST` writes elsewhere in this crate.
        let response = crate::http::send(&self.http, &self.retry, true, |c| {
            c.put(format!("{}/me/presence", self.server_url))
                .bearer_auth(&self.token)
                .json(&UpdatePresenceRequest { status })
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        Ok(())
    }

    /// `PUT /presence/{identity_id}` (closed out by #741/#749) — an
    /// *integrator* publishing presence on behalf of this session's own
    /// identity, distinct from [`Session::update_presence`]'s own
    /// self-service path: this one requires this integrator's challenge-
    /// response credential plus an active `presence.publish` grant (the
    /// same `x-avalon-integrator-*`/`x-avalon-identity-id` header shape
    /// `achievements::Session::submit_achievement_issuance` already uses
    /// for an integrator acting on this session's identity), and can set
    /// `active_in` — which the self-service path deliberately cannot.
    /// `active_in`, if set at all, must be this integrator's own id
    /// (`AvalonConfig::integrator_slug`'s resolved integrator) — checked
    /// server-side (`AppError::PresenceActiveInMismatch`), not merely
    /// assumed client-side.
    pub async fn update_integrator_presence(
        &self,
        status: PresenceStatus,
        active_in: Option<Uuid>,
    ) -> Result<(), SdkError> {
        let slug = self
            .integrator_slug
            .as_deref()
            .ok_or(SdkError::MissingIssuerCredentials)?;
        let signing_key_bytes = self.signing_key.ok_or(SdkError::MissingIssuerCredentials)?;
        let signing_key = SigningKey::from_bytes(&signing_key_bytes);

        #[derive(Deserialize)]
        struct ChallengeResponse {
            challenge_id: Uuid,
            nonce: String,
        }
        #[derive(Serialize)]
        struct UpdateIntegratorPresenceRequest {
            status: PresenceStatus,
            #[serde(skip_serializing_if = "Option::is_none")]
            active_in: Option<Uuid>,
        }

        let challenge_response = crate::http::send(&self.http, &self.retry, false, |c| {
            c.post(format!("{}/integrations/{slug}/challenge", self.server_url))
        })
        .await?;
        if !challenge_response.status().is_success() {
            return Err(crate::http::map_error_response(challenge_response).await);
        }
        let challenge: ChallengeResponse = challenge_response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        let nonce = BASE64
            .decode(&challenge.nonce)
            .map_err(|_| SdkError::MissingIssuerCredentials)?;
        let challenge_signature = signing_key.sign(&nonce);

        let subject = self.identity().id.0;
        let response = crate::http::send(&self.http, &self.retry, false, |c| {
            c.put(format!("{}/presence/{subject}", self.server_url))
                .header("x-avalon-integrator-key-id", &self.integrator_key_id)
                .header(
                    "x-avalon-integrator-challenge-id",
                    challenge.challenge_id.to_string(),
                )
                .header(
                    "x-avalon-integrator-signature",
                    BASE64.encode(challenge_signature.to_bytes()),
                )
                .header("x-avalon-identity-id", subject.to_string())
                .json(&UpdateIntegratorPresenceRequest { status, active_in })
        })
        .await?;
        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }
        Ok(())
    }

    /// Subscribes to live presence updates for `ids` — issue #136, additive
    /// to `presence_of`'s point-in-time reads, not a replacement for them.
    /// Requires `presence.read`, same as every other presence method here.
    ///
    /// Connects to `GET /ws/presence` (auth via a `?token=` query parameter
    /// — a websocket handshake can't carry a bearer header, see
    /// `crates/server/src/handlers.rs::authenticate_token`'s own doc
    /// comment), sends one `subscribe` message for `ids`, then spawns a
    /// background task forwarding every [`Presence`] the server pushes into
    /// the returned channel. Dropping the receiver drops the sender on the
    /// task's next send attempt, which ends the task — no separate
    /// `unsubscribe` call is needed.
    pub async fn subscribe_presence(
        &self,
        ids: &[IdentityId],
    ) -> Result<tokio::sync::mpsc::UnboundedReceiver<Presence>, SdkError> {
        self.require(Capability::PresenceRead)?;

        let url = websocket_url(
            &self.server_url,
            &format!("/ws/presence?token={}", self.token),
        );
        let (ws_stream, _) = tokio_tungstenite::connect_async(&url)
            .await
            .map_err(|e| SdkError::WebSocket(e.to_string()))?;
        let (mut write, mut read) = ws_stream.split();

        let subscribe =
            serde_json::to_string(&PresenceSubscribeMessage::Subscribe { ids: ids.to_vec() })
                .expect("PresenceSubscribeMessage always serializes");
        write
            .send(WsMessage::Text(subscribe.into()))
            .await
            .map_err(|e| SdkError::WebSocket(e.to_string()))?;

        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
        tokio::spawn(async move {
            while let Some(Ok(message)) = read.next().await {
                let WsMessage::Text(text) = message else {
                    continue;
                };
                let Ok(presence) = serde_json::from_str::<Presence>(&text) else {
                    continue;
                };
                if tx.send(presence).is_err() {
                    break;
                }
            }
        });

        Ok(rx)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::identity::{Identity, Profile};
    use time::OffsetDateTime;
    use uuid::Uuid;

    fn make_friendship(a: Uuid, b: Uuid) -> Friendship {
        Friendship {
            a: IdentityId(a),
            b: IdentityId(b),
            since: OffsetDateTime::now_utc(),
        }
    }

    fn make_presence(identity_id: IdentityId) -> Presence {
        Presence {
            identity_id,
            status: PresenceStatus::Online,
            active_in: None,
            updated_at: OffsetDateTime::now_utc(),
        }
    }

    /// Builds a `Session` with no live server behind it — every test here
    /// either never makes a request (capability check fails first) or
    /// exercises pure logic (`merge_friend`) directly.
    fn test_session(granted: Vec<&str>) -> Session {
        let self_id = IdentityId(Uuid::new_v4());
        Session {
            identity: Identity {
                id: self_id,
                created_at: OffsetDateTime::now_utc(),
            },
            profile: Profile {
                identity_id: self_id,
                display_name: "test".to_string(),
                avatar_url: None,
                bio: None,
                favorite_genres: Vec::new(),
                pronouns: None,
                banner_url: None,
                status: None,
                links: Vec::new(),
                timezone: None,
                theme_color: None,
                location: None,
                main_guild: None,
            },
            granted: granted.into_iter().map(Capability::from).collect(),
            http: reqwest::Client::new(),
            // Deliberately unroutable — these tests must never actually
            // reach the network; an attempted connection here would hang or
            // error in a way that's obviously not `CapabilityNotGranted`.
            server_url: "http://127.0.0.1:1".to_string(),
            token: "test-token".to_string(),
            integrator_key_id: "test-key".to_string(),
            integrator_slug: None,
            signing_key: None,
            retry: crate::RetryConfig::default(),
        }
    }

    #[tokio::test]
    async fn friends_without_grant_is_rejected_before_any_request() {
        let session = test_session(vec![]);
        let result = session.friends().await;
        assert!(matches!(result, Err(SdkError::CapabilityNotGranted(_))));
    }

    #[tokio::test]
    async fn presence_without_grant_is_rejected_before_any_request() {
        let session = test_session(vec![]);
        let result = session.presence().await;
        assert!(matches!(result, Err(SdkError::CapabilityNotGranted(_))));
    }

    #[tokio::test]
    async fn presence_of_without_grant_is_rejected_before_any_request() {
        let session = test_session(vec![]);
        let result = session.presence_of(&[IdentityId(Uuid::new_v4())]).await;
        assert!(matches!(result, Err(SdkError::CapabilityNotGranted(_))));
    }

    #[tokio::test]
    async fn presence_of_with_no_ids_short_circuits_before_any_request() {
        let session = test_session(vec!["presence.read"]);
        let result = session.presence_of(&[]).await;
        assert!(result.unwrap().is_empty());
    }

    #[test]
    fn merge_friend_has_no_presence_when_map_is_empty() {
        let self_id = IdentityId(Uuid::new_v4());
        let other_id = IdentityId(Uuid::new_v4());
        let friendship = make_friendship(self_id.0, other_id.0);

        let friend = merge_friend(&friendship, self_id, &HashMap::new());

        assert_eq!(friend.identity_id, other_id);
        assert!(friend.presence.is_none());
    }

    #[test]
    fn merge_friend_picks_up_presence_when_present_in_map() {
        let self_id = IdentityId(Uuid::new_v4());
        let other_id = IdentityId(Uuid::new_v4());
        let friendship = make_friendship(self_id.0, other_id.0);
        let mut presence_by_id = HashMap::new();
        presence_by_id.insert(other_id, make_presence(other_id));

        let friend = merge_friend(&friendship, self_id, &presence_by_id);

        assert_eq!(friend.identity_id, other_id);
        assert!(friend.presence.is_some());
    }

    #[test]
    fn merge_friend_resolves_the_other_party_regardless_of_a_b_order() {
        let self_id = IdentityId(Uuid::new_v4());
        let other_id = IdentityId(Uuid::new_v4());

        // self as `a`
        let friendship_1 = make_friendship(self_id.0, other_id.0);
        assert_eq!(
            merge_friend(&friendship_1, self_id, &HashMap::new()).identity_id,
            other_id
        );

        // self as `b`
        let friendship_2 = make_friendship(other_id.0, self_id.0);
        assert_eq!(
            merge_friend(&friendship_2, self_id, &HashMap::new()).identity_id,
            other_id
        );
    }

    #[tokio::test]
    async fn subscribe_presence_without_grant_is_rejected_before_any_connection() {
        let session = test_session(vec![]);
        let result = session
            .subscribe_presence(&[IdentityId(Uuid::new_v4())])
            .await;
        assert!(matches!(result, Err(SdkError::CapabilityNotGranted(_))));
    }
}
